### PR TITLE
Emit watch changes as a plugin hook

### PR DESCRIPTION
**Pull request by guybedford**
This PR contains:

- [ ] bugfix
- [x] feature
- [ ] refactor
- [ ] tests
- [ ] documentation
- [ ] metadata

### Breaking Changes?

- [ ] yes
- [x] no

### Please Describe Your Changes

Fixes https://github.com/rollup/rollup/issues/2383 adding a watch hook for change events.

The idea here as well is to then deprecate the `this.watcher` as we should now have all the info we need here.

### CODE REVIEW

File: src/rollup/types.d.ts [hunk 1]
```diff
@@ -91,12 +91,14 @@ export interface PluginCache {
...
 
 export interface PluginContext {
+	// deprecate:
 	watcher: Watcher;
 	addWatchFile: (id: string) => void;
 	cache: PluginCache;
 	resolveId: ResolveIdHook;
 	isExternal: IsExternal;
 	parse: (input: string, options: any) => ESTree.Program;
+	// deprecate asset dependencies
```
Comment (tivac) [inline src/rollup/types.d.ts]: I didn't even know that assets could have dependencies and it's already being deprecated 🤔😞

**Comment by tivac** [comment]
This'll be a big help, thanks so much @guybedford!

File: src/rollup/types.d.ts [hunk 2]
```diff
@@ -203,6 +205,7 @@ export interface Plugin {
 	footer?: AddonHook;
 	intro?: AddonHook;
 	outro?: AddonHook;
+	watchChange?: (id: string) => void;
```
Comment (lukastaegert) [inline src/rollup/types.d.ts]: So the idea is to deprecate asset dependencies in favour of the `watchChange` handler? Sounds really good to me, much clearer and easier to reason about.

In order to start the actual deprecation soon, should we
* Remove the "dependencies" version `emitAssets` from the types, or at least
* Add an [@deprecated](http://usejsdoc.org/tags-deprecated.html) JSDoc comment above the type which should be picked up by many IDEs? (would be an interesting thing to do for other deprecations as well. Not 100% sure about VSCode but I know WebStorm/IntelliJ uses these `@deprecated` tags to visually mark usages as deprecated).

Also I agree with @shellscape that a warning would make sense considering asset dependencies are not widely known yet anyway (thanks @tivac for pointing that out 😉)

**Comment by guybedford** [comment]
I was going to do a PR specifically just on the deprecations for 1.0 (adding warnings, but not actually making them breaking).

This asset dependencies one could be one we deprecate before 1.0 though potentially, as I don't think anyone is using this feature yet as it's been very early for this.

I could add the removal here if you like?

**Comment by guybedford** [comment]
Asset dependencies were not documented for stability reasons (I like to give all new APIs a grace period personally), so removing it should be fine actually.

**Comment by guybedford** [comment]
I've added the asset dependencies removal.

**Pull request merged**